app/models/ems_refresh/parsers/azure.rb

```diff
@@ -0,0 +1,214 @@
+module EmsRefresh
+  module Parsers
+    class Azure < ManageIQ::Providers::CloudManager::RefreshParser
+      VALID_LOCATION = /\w+/i
+
+      def self.ems_inv_to_hashes(ems, options = nil)
+        new(ems, options).ems_inv_to_hashes
+      end
+
+      def initialize(ems, options = nil)
+        @ems        = ems
+        ems.connect
+        @vmm        = ::Azure::Armrest::VirtualMachineManager.new
+        @asm        = ::Azure::Armrest::AvailabilitySetManager.new
+        @options    = options || {}
+        @data       = {}
+        @data_index = {}
+      end
+
+      def ems_inv_to_hashes
+        log_header = "MIQ(#{self.class.name}.#{__method__}) Collecting data for EMS : [#{@ems.name}] id: [#{@ems.id}]"
+
+        $log.info("#{log_header}...")
+        get_series
+        get_availability_sets
+        get_vms
+        clean_up_extra_flavor_keys
+        $log.info("#{log_header}...Complete")
+
+        @data
+      end
+
+      private
+
+      def get_series
+        series = []
+        get_locations.each do |location|
+          begin
+            series << @vmm.series(location)
+          rescue RestClient::BadGateway
+            next
+          end
+        end
+        series = series.flatten!.uniq
```

Comment:
Don't use the `!` method in the middle because if it doesn't actually flatten anything it will return nil and you will call .uniq on nil.